### ISSUE TITLE
DESIGN: State space: default standard errors: OPG vs OIM

### ISSUE THREAD

**Issue by ChadFulton** [statsmodels/statsmodels#5466]
**TL;DR**: we currently use outer-product of gradients (OPG) as the default information matrix approximation, as does Stata and Eviews.  Some limited experiments on AR, MA models (see below) suggest that this may not be the best option. Does anyone know of results / references / reasons why we should use OPG over one of the other methods? Is it just the guaranteed nonsingularity of the resulting matrix?

----------

Currently the default in state space models is to approximate the information matrix via the outer product of gradients (OPG).  This is e.g. also the default in Stata and Eviews for state space models and for arima models.

However, I'm not aware a reason / reference that makes this the preferred option for approximating the information matrix, other than the fact that the OPG itself is guaranteed to be positive (semi)definite?

Both Stata and Eviews have the observed information matrix (OIM) computation available, although it is not the default. I'm wondering if anyone knows why this is, given that it appears (see below) that the OIM may do a better job, and even that the numerical approximation to the Hessian may do a better job.

Since Brockwell and Davis (2002) give some analytic formulas for the covariance of AR(p) and MA(q) 
models for p=1,2 and q=1,2, I thought it might be interesting to compare the three methods for approximating the information matrix, available in state space models:

1. Outer product of gradients (computed using `score_obs`)
2. Numerical approximation of the Hessian (via complex step differentiation)
3. Observed information matrix (computation described in Harvey, 1989)

The results of my experiment are at: https://gist.github.com/ChadFulton/986f7bfbf90e162e8f3526880a833891

It appears (in the small number of examples that I looked at) that the OIM gives the closest answer to the theoretical covariance matrix, and that the numerical approximation is often as good or only slightly worse. In these examples, the OPG is always worst.

**Comment by josef-pkt** [comment]
I don't know anything about your usecase.
In GLM we added an option to use EIM (same as OPG, IIRC) instead of OIM during the optimization because it is faster. However, the final cov_params are not affected by this and uses OIM/EIM depending on the algorithm, but uses OIM/Hessian in general.

I think Newton with OPG corresponds to BHHH, but I never verified this.

The main reason to avoid the full Hessian in GLM, besides numerical problems that might loose positive definiteness when away from the optimum, is that computing the OIM/Hessian is much more expensive than EIM/OPG in large datasets (e.g. for 5 million observations).

(Hessian in GLM is based on analytic derivatives, but those have many extra terms in general which makes it computationally expensive.)

Aside: mean-variance separability:
In all current cases (OLS/GLM/RLM) we use the property that mean and variance estimated are asymptotically uncorrelated or independent, i.e. the Hessian is only for the mean term. Implicitly we exploit that the expected or asymptotic hessian/OIM is block diagonal in mean-variance for the linear exponential family (LEF). 
So it's not a full Hessian including auxiliary parameter like scale.
This will not be correct in models that don't have a LEF or a symmetric distribution.

Another aside: robustness to noise or outliers in small samples
Accurately matching a theoretical asymptotic cov_params might not be the best choice in small samples. A simplified approximation might be better than an asymptotically accurate cov_params in small samples if it is less sensitive to noise and outliers. 
popular example: sandwich cov_params can become very noisy in small samples compared to an asymptotically incorrect "naive" cov_params.
another one: estimating 3rd and 4th moment for a general (fewer assumptions) cov_params, can be worse than imposing moments under normality assumptions which only requires estimation of first and second moments (even if normality assumption is a bit misspecified).

In the GLM literature there is some preference for OIM compared to EIM/OPG for inference because it is more robust to some deviations from the assumptions of a correctly specified likelihood.

**Comment by ChadFulton** [comment]
Thanks Josef, that is very useful / interesting!

**Comment by thequackdaddy** [comment]
> The main reason to avoid the full Hessian in GLM, besides numerical problems that might loose positive definiteness when away from the optimum, is that computing the OIM/Hessian is much more expensive than EIM/OPG in large datasets (e.g. for 5 million observations).

So for me, I've found that on my large-ish datasets of 5-10 you sometimes get faster optimization with OIM and sometimes with EIM. Each iteration with OIM is a little longer... maybe 20-40% more time. However, OIM sometimes can fit in fewer iterations. Which one is faster really seems to be a wash.
